kie-server-parent/kie-server-services/kie-server-services-openshift/src/main/java/org/kie/server/services/openshift/impl/storage/cloud/KieServerStateOpenShiftRepository.java

```diff
@@ -175,37 +186,128 @@ public KieServerState load(@NotNull String serverId) {
     
     @Override
     public boolean isKieServerReady() {
-        if (isKieServerRuntime()) {
-            return KieServerLocator.getInstance().isKieServerReady();
-        } else {
-            /**
-             * For non KieServer environment, i.e. Workbench, it assumes KieServer is ready.
-             */
-            return true;
-        }
+        return KieServerLocator.getInstance().isKieServerReady();
     }
 
     /**
-     * To provide compatibility to non kie server use case, such as supporting Workbench or 
-     * standalone kie server controller, this utility method indicates if the runtime environment
-     * is kie server or not.
+     * To be compatible with non kieserver process, (workbench), kie container at certain status,
+     * i.e. STOPPED, should not be removed.
+     * @param dc
+     * @param c
+     * @param newState
      * @return
      */
-    private boolean isKieServerRuntime() {
-        return System.getProperty(KieServerConstants.KIE_SERVER_ID) != null;
+    public boolean isKieContainerRemovalAllowed(ConfigMap cm, KieServerState newState) {
+        KieServerState state = (KieServerState) xs.fromXML(cm.getData().get(CFG_MAP_DATA_KEY));
+        for (KieContainerResource container : state.getContainers()) {
+            if (container.getStatus().equals(KieContainerStatus.STOPPED) &&
+                newState.getContainers().stream()
+                    .noneMatch(c -> c.getContainerId().equals(container.getContainerId()))) {
+                // STOPPED kie container not found
+                logger.warn("Removing STOPPED KieContainer by KieServer process is not allowed.");
+                return false;
+            }
+        }
+        return true;
+    }
+    
+    /**
+     * Non KieServer process including workbench is allowed to update KieServerState
+     * even during DC rollout, but ONLY under certain conditions.
+     * @param cm
+     * @param newState
+     * @return
+     */
+    public boolean isKieContainerUpdateDuringRolloutAllowed(ConfigMap cm, KieServerState newState) {
+        KieServerState state = (KieServerState) xs.fromXML(cm.getData().get(CFG_MAP_DATA_KEY));
+        for (KieContainerResource container : state.getContainers()) {
+            if (container.getStatus().equals(KieContainerStatus.STARTED) &&
+                !newState.getContainers().stream()
+                         .noneMatch(c -> c.getContainerId().equals(container.getContainerId()) &&
```

Comment:
Could you please use anyMatch() here and remove negation?
I think it would improve readability :)

Reply:
Good point, changed as per suggestion.